ancestrymanager/ancestrymanager.go

```diff
@@ -1,189 +1,263 @@
-	if err != nil {
-		return "", err
+func (m *manager) getAncestorsWithCache(key string) ([]string, error) {
+	var ancestors []string
+	cur := key
+	for cur != "" {
+		if cachedAncestors, ok := m.ancestorCache[cur]; ok {
+			ancestors = append(ancestors, cachedAncestors...)
+			break
+		}
+		if strings.HasPrefix(cur, "organizations/") {
+			ancestors = append(ancestors, cur)
+			break
+		}
+		if m.resourceManager == nil {
+			return nil, fmt.Errorf("resourceManager required to fetch ancestry for %s from the API", cur)
+		}
+		project, err := m.resourceManager.Projects.Get(cur).Do()
+		if err != nil {
+			return nil, err
+		}
+		ancestors = append(ancestors, project.Name)
+		cur = project.Parent
 	}
-	path := ancestryPath(ancestry.Ancestor)
-	m.store(project, path)
-	return path, nil
+	m.store(key, ancestors)
+	return ancestors, nil
 }
 
-func (m *manager) store(project, ancestry string) {
-	if project != "" && ancestry != "" {
-		m.ancestryCache[project] = ancestry
+func (m *manager) store(key string, ancestors []string) {
+	if key == "" || len(ancestors) == 0 {
+		return
 	}
-}
-
-// GetAncestryWithResource first attempts to get Ancestry from the resource
-// If that fails, it falls back to the offline cache.
-func (m *manager) GetAncestryWithResource(project string, tfData resources.TerraformResourceData, cai resources.Asset) (string, error) {
-	ancestry, ok := m.getAncestryFromResource(tfData, cai)
-	if ok {
-		path := ancestryPath(ancestry)
-		m.store(project, path)
-		m.errorLogger.Info(fmt.Sprintf("Retrieved ancestry for %s: %s", project, path))
-		return path, nil
+	if _, ok := m.ancestorCache[key]; !ok {
+		m.ancestorCache[key] = ancestors
 	}
-
-	path, err := m.GetAncestry(project)
-	if err != nil {
-		return "", err
+	// cache ancestors along the ancestry path
+	for i, ancestor := range ancestors {
+		if _, ok := m.ancestorCache[ancestor]; !ok {
+			m.ancestorCache[ancestor] = ancestors[i:]
+		}
 	}
-
-	return path, nil
 }
 
-// New returns AncestryManager that can be used to fetch ancestry information for a project.
-// entries takes project as key and ancestry as value to pre-warm the offline cache. If no
-// resourceManager is provided, API requests for ancestry will be disabled.
-func New(resourceManager *cloudresourcemanager.Service, entries map[string]string, errorLogger *zap.Logger) (AncestryManager, error) {
-	am := &manager{
-		ancestryCache: map[string]string{},
-		resourceAncestryManager: resourceAncestryManager{
-			errorLogger: errorLogger,
-		},
-		resourceManager: resourceManager,
+func parseAncestryPath(path string) ([]string, error) {
+	normStr := normalizeAncestry(path)
+	splits := strings.Split(normStr, "/")
+	if len(splits)%2 != 0 {
+		return nil, fmt.Errorf("unexpected format of ancestry path %s", path)
 	}
-	for project, ancestry := range entries {
-		if ancestry != "" {
-			ancestry = fmt.Sprintf("%s/project/%s", ancestry, project)
+	var ancestors []string
+	for i := len(splits) - 1; i >= 0; i = i - 2 {
```

Comment:
not a blocker but it seems like this might be a little easier to read if `i` was the index of the front part and the loop went forward through the ancestry - something like:

```suggestion
	allowedPrefixes := map[string]bool{
		"projects": true,
		"folders": true,
		"organizations": true,
	}
	for i := 0; i < len(splits); i = i + 2 {
		if _, ok := allowedPrefixes[splits[i]]; !ok {
			return nil, fmt.Errorf("invalid ancestry path %s with %s", path, splits[i])
		}
		ancestors = append(ancestors, fmt.Sprintf("%s/%s", splits[i], splits[i+1]))
	}
```

Reply:
Done. It had to reverse the sequence at the end though, if i start from 0.
I added the reverse at the end.